[12.0] contract: some refactoring and a UX improvement

sbidoul opened the pull request:
Some simple refactoring of the contract module. The main idea is to split the computation of the next invoice date in two parts: the computation of the next period, followed by the computation of the next invoice date (aka `recurring_next_date`) based on the next period dates.

This is best reviewed one commit at a time to understand individual changes.

There is also a UX improvement: display the next period start and end date on the contract line form, so the user does not need to deduce it from other parameters.

Finally, `monthlylastday` now supports `pre-paid`/`post-paid`. This could be a breaking change
if there is code that creates `monthlylastday` contract lines without setting them to `post-paid` mode.

fixes: #428

sbidoul committed:
[REF] contract: rename misnamed methods

sbidoul committed:
[REF] contract: clarify _get_recurring_next_date

First compute the next period end date,
then derive the next invoice date from the next
period stard and end date.

sbidoul committed:
[REF] contract: handle max_date_end in _get_recurring_next_date

This concentrates all next date calculation
logic in one place, and will allow further simplifications.

sbidoul committed:
[IMP] contract: add next period start/end fields

Add two computed field showing the next period
start and end date. This improve the UX and will
enable further simplifications in the code.

sbidoul committed:
[REF] contract: refactor _get_period_to_invoice

Move the part of the logic that compute the next
period depending on the chosen next invoice date
to _get_next_period_date_end.

sbidoul committed:
[REF] contract: refactor _update_recurring_next_date

Reuse the logic that is now fully located in _get_recurring_next_date.

sbidoul committed:
[REF] contract: re-add _compute_first_recurring_next_date

For backward compatibility

sbidoul committed:
[FIX] contract: add missing dependency in computed field

sbejaoui commented on contract/tests/test_contract.py:
```diff
@@ -279,7 +279,7 @@ def test_last_invoice_post_paid(self):
         )
         self.contract.recurring_create_invoice()
         self.assertEqual(
-            self.acct_line.recurring_next_date, to_date('2018-04-01')
```
The current version assure that a customer is billed the same day each month.

Assume this example:
- date start 01/01
- date end 15/03

the customer will receive an invoices at:
* for pre-paid:  01/01, 01/02 and 01/03
* for monthlylastday:  31/01, 28/02 and 31/03
* for post-paid: 01/02, 01/03 and 01/04

for me it make more sens to say post-paid mode invoice date is the last day of the invoicing period (or the last day of the invoicing period + one day).

Let make a RFC and see what other think about this change

sbejaoui commented on contract/tests/test_contract.py:
```diff
@@ -247,7 +247,7 @@ def test_contract_yearly_pre_paid(self):
 
     def test_contract_monthly_lastday(self):
-        recurring_next_date = to_date('2018-03-31')
+        recurring_next_date = to_date('2018-02-28')
```
+1

sbidoul committed:
[REF] contract: remove one monthlylastday special case

get_relative_delta now works the same for all recurring rules.
Move the special case handling to _init_last_date_invoiced
which is used only for migration.

sbidoul committed:
[IMP] contract: support pre-paid for monthlylastday

monthlylastday is (almost) not a special case anymore \o/.
montlylastday is simply a montly period where the
periods are aligned on month boundaries.
The last bit of special casing is that postpaid generates
invoice the day after the last dasy of the period, except
for monthlylastday where the invoice is generated on the
last day of the period. This last exception will disappear
when we put the offset under user control.

This is a breaking change because the post-paid/pre-paid
mode becomes relevant for monthlylastday invoicing.
The field becomes visible in the UI. Code that generate
monthlylastday contract lines must now correctly set
the pre-paid/post-paid mode too. Some tests have had
to be adapted to reflect that.

sbejaoui commented on contract/migrations/12.0.5.0.0/pre-migration.py:
```diff
@@ -0,0 +1,10 @@
+    # make sure it has the value that was implied for previous versions.
+    cr.execute(
+        """\
+            UPDATE contract_contract
```
```suggestion
            UPDATE contract_line
```

sbidoul committed:
[REF] contract: make recurring_invoicing_offset a computed field

In preparation to making it user modifiable.

sbejaoui committed:
[FIX] contract: Improve unit tests

sbidoul committed:
[REF] contract: make get_next_period_date_end public

Make it public because it is the core logic of the module.
Also, clarify that recurring_invoicing_type
and recurring_invoicing_offset are needed only when
we want the next period to be computed from a
user chosen next invoice date.

sbidoul committed:
[REF] contract: rename _get_recurring_next_date as get_next_invoice_date

It is easier to understand. Also make it public.

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -664,7 +664,7 @@ def _get_period_to_invoice(
         first_date_invoiced = False
         if not recurring_next_date:
-            return first_date_invoiced, last_date_invoiced, recurring_next_date
+            return first_date_invoiced, False, recurring_next_date
```
It would be more readable to `return False, False, False`

sbejaoui replied:
yes

sbejaoui replied:
```suggestion
            return False, False, False
```

sbejaoui committed:
[IMP] contract: add unit test for different combinations for next invoicing period

sbejaoui reviewed (approved):
Fantastic

Thank you @sbidoul for this great work.

For me the most important thing in this PR is the fact that it fix a logic issue we had for a long time. Now the `recurring_invoicing_type` is no longer implicit in the `monthlylastday` option and we can configure a invoicing process to stop at the end of the month and be payed at the first day of the period (even for an interval of several months [see](https://github.com/OCA/contract/blob/eef75b7d88fef7d4385bce9b137b85e7e94381b5/contract/tests/test_contract.py#L973)).

Adding to this fix, this PR open the way for new features (invoicing offset, computed next invoice date,.. ) by simplifying the code and improve the UX.

cc/ @gva-acsone

sbejaoui committed:
[REF] contract: simplify _get_period_to_invoice

sbidoul commented on contract/models/contract_line.py:
```diff
@@ -669,7 +669,7 @@ def _get_period_to_invoice(
         self.ensure_one()
         first_date_invoiced = False
```
This line is useless now

sbidoul committed:
[IMP] contract: display invoicing offset

sbidoul commented:
I added the offset field in the contract line form, to further help the user understand what is going on.

@sbejaoui will do some more testing with other depending modules (OCA and customer specific), but we don't expect issue. So this is ready to review. Reviewing individual commits is highly recommended to understand the logical evolution of the code base.

sbejaoui committed:
[IMP] - Set recurring_invoicing_type for existing products set to monthlylastday

rafaelbn reviewed (approved):
After a first quick functional review it looks OK :+1:

pedrobaeza commented:
It's difficult to predict side effects with the great amount of changes, but let's trust in your good work and extra tests also makes this more confidence. Trying it out in real will tell us any possible problem.

/ocabot merge major

sbidoul commented:
To summarize breaking changes:

- Contract lines and products using `monthlylastday` MUST now specify `pre-paid` / `post-paid`. There is a migration script for that. The UI has been adapted. Code that creates such lines must be adapted or will default to `pre-paid` whereas `post-paid` was implied before.
- `get_relative_delta()` now works uniformly for all period types (it had an exception for `monthlylastday` before). Code calling it directly must be adapted.
- In `post-paid` mode, the invoice date is now *always* the first day after the end of the period being invoiced. There was an exception before for the last period of the contract.
